### PR TITLE
feat: Support overriding native endorsed plugins

### LINKED ISSUE #59657: Support overriding the implementation of a federated plugin
## Use case

For devices, without GMS we want to override the default implementation of packages like `firebase_auth`.
To achieve this we want to implement the `firebase_auth_platform_interface` and use our custom class.

## Proposal

In the federated plugin proposal by @amirh, there is a section about overriding the default implementation with a custom package:
https://docs.google.com/document/d/1LD7QjmzJZLCopUrFAAE98wOUQpjmguyGTN2wd_89Srs/edit#heading=h.zc9ie62xv82k

We thought something like this should be possible:

pubspec.yaml
```
dependencies: 
  firebase_auth: any
  firebase_auth_android_rest: any
```

And because the Flutter tool knows that `firebase_auth_android_rest` provides a custom implementation of the `firebase_auth_platform_interface`/`firebase_auth` package, it uses `firebase_auth_android_rest` and not `firebase_auth` for Android.

We don't know if this is possible yet.

### LINKED ISSUE #80374: [Federated Plugins] Provide a way to override a default/endorsed platform implementation
## Use case

It would be desirable for a federated plugin to have a default implementation for iOS, either inside the plugin or an endorsed separate package, and to then allow apps to choose an alternative platform implementation to override that default. E.g. for just_audio, we will have a default AVQueuePlayer-based implementation, but an alternative AudioKit implementation and an alternative VLC implementation.

However, I don't think this override ability has been added to Flutter yet, although it was implied in the [design doc](https://docs.google.com/document/d/1LD7QjmzJZLCopUrFAAE98wOUQpjmguyGTN2wd_89Srs/edit#).

What happens instead is that the generated plugin registrant will automatically add both the default/endorsed plugin AND any additional plugin the app adds to their pubspec with the intention of overriding the default. Since both are listed, alphabetical order will dictate which one is effective.

## Proposal

Currently a platform implementation is endorsed by adding its dependency to the plugin's front-end and so it's going to get added to the generated plugin registrant regardless of any attempt to override it.

I'm not sure on the best solution, but maybe either the mechanism to endorse an implementation needs to be something other than a normal dependency, or otherwise pub should add some special cases to the code that generates the generated plugin registrant to detect overrides and avoids adding both the override and the default.

### PR DESCRIPTION
**Pull request by Gustl22**
These changes allow to override existing native endorsed (federated & inline) plugins with e.g. non-endorsed plugin implementations via direct dependencies as described in the section *Platform implementation selection* in the [design doc](https://docs.google.com/document/d/1LD7QjmzJZLCopUrFAAE98wOUQpjmguyGTN2wd_89Srs).

```yaml
# pubspec.yaml
name: example
dependencies:
  my_plugin: ^0.0.1
  my_plugin_android_alternative: ^0.0.1
```

Closes #80374, closes #59657

Related: Override Dart plugins (#87991, #79669)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.
- [x] Move to own PR for `.flutter-plugins` migration, see #48918

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by Gustl22** [comment]
Blocked by #137115

**Comment by zr6328** [comment]
My pubspec.yaml is as follows:
  audioplayers: ^1.0.0  
  custom_audio_player:
    git:
      url: xxx
      path: xxx
      ref: xxx
I use pure dart to implement the audioplayers. But when I use the flutter build linux, the registerWith function is not called in the file dart_plugin_registrant.dart

**Comment by zr6328** [comment]
# pubspec.yaml
name: example
dependencies:
  audioplayer: ^0.0.1
  myAudioPlayer: ^0.0.1

**Comment by zr6328** [comment]
`dependencies: 
      audioplayers: ^1.0.0
      my_audio_player: ^1.0.0 
`
the my_audio_player plugin is created based on audioplayers, the pubspec.yaml is show as follows:
`flutter:
      plugin:
           implements: audioplayers
           platforms:
               linux:
                    dartPluginClass: AudioPlayerPlugin
`

when i run the flutter build linux, the function of registerWith in the dart_plugin_registrant.dart is not called. Please give me some methods to solve this , thanks a lot

**Comment by stuartmorgan** [comment]
It's hard to see what's changing here since it's mixed with some large-change refactors (most notably extracting `_getPossiblePluginResolutions`); pulling any purely code refactor into a simple prequel PR that doesn't also change logic would be useful in getting feedback here.

**Comment by Gustl22** [comment]
Thanks, yes that was also my intention, I will have time in the coming week to work on it :)

**Comment by Gustl22** [comment]
@stuartmorgan Finally ready for review. Best regards.

File: packages/flutter_tools/lib/src/flutter_plugins.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/flutter_plugins.dart:1147]: `selectDartPluginsOnly: false` sounds like it would include everything, including Dart plugins. If that's not how it works, it should be `selectDartPlugins`. Or it could be a file-private enum where the two types are `dart` and `native`, and the paramater could be `pluginTypes`.

File: packages/flutter_tools/lib/src/flutter_plugins.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/flutter_plugins.dart:1207]: Dart-only

File: packages/flutter_tools/test/general.shard/dart_plugin_test.dart
Comment (stuartmorgan) [inline packages/flutter_tools/test/general.shard/dart_plugin_test.dart:495]: Why is this test being removed?

File: packages/flutter_tools/test/general.shard/plugins_test.dart
Comment (stuartmorgan) [inline packages/flutter_tools/test/general.shard/plugins_test.dart:1133]: How about "user-selected implementation overrides inline implementation"

File: packages/flutter_tools/test/general.shard/plugins_test.dart
Comment (stuartmorgan) [inline packages/flutter_tools/test/general.shard/plugins_test.dart:1238]: There should also be a "user-selected implementation overrides default implementation" test.

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
Mostly looks good, just minor comments. Thanks for all the refactoring that made this PR so clear!

File: packages/flutter_tools/test/general.shard/dart_plugin_test.dart
Comment (Gustl22) [reply packages/flutter_tools/test/general.shard/dart_plugin_test.dart:495]: It just looks like it was removed, because of the git diff is not showing it correctly. It is present here: https://github.com/flutter/flutter/pull/137040/files#diff-2fb2f7fe1d44f737a3b8e0fefdf353e06196290b8461daf7e2b7c33cb615f6a5R582
In the code editor it is much more clear that only the formatting and the `selectDartPluginsOnly: true` flag has changed for all the tests.

File: packages/flutter_tools/test/general.shard/plugins_test.dart
Comment (Gustl22) [reply packages/flutter_tools/test/general.shard/plugins_test.dart:1133]: I wanted to conform the description in the `dart_plugin_test.dart`, where the same is tested, but I can change the it for both then. Makes sense to have the keyword `override` in it.

File: packages/flutter_tools/test/general.shard/plugins_test.dart
Comment (Gustl22) [reply packages/flutter_tools/test/general.shard/plugins_test.dart:1238]: The test for it can be found [here](https://github.com/flutter/flutter/pull/137040/files/37ce21a454de466e589ff96cd39684a46c4f4029#diff-4b59d4ff0ace84a3aaa79567f0588a1bb72a8cfb72558a9e1548ff33b0de8ffbR1181).

I suspect you meant a test for `Build time plugin injection` aka **web**. I added a test for it to ensure the injection code is also working for web in `injectBuildTimePluginFiles` in `flutter_plugins.dart`. The plugin resolution code is shared with the native code, which is already covers this case. But if it feels more safe, I can add this test, too (?)

File: packages/flutter_tools/test/general.shard/plugins_test.dart
Comment (Gustl22) [reply packages/flutter_tools/test/general.shard/plugins_test.dart]: Done in e637632102c53518f11babf7e77dd2cb7d21c510 .

File: packages/flutter_tools/lib/src/flutter_plugins.dart
Comment (Gustl22) [reply packages/flutter_tools/lib/src/flutter_plugins.dart]: Done in f2a85fdfc2dc0e690bf03256bcda960e20067d86

File: packages/flutter_tools/lib/src/flutter_plugins.dart
Comment (Gustl22) [reply packages/flutter_tools/lib/src/flutter_plugins.dart:1153]: I don't wanted to expose the enum, so I used a bool for the public methods and the enum for the internal handling as the code is then much more readable indeed. Done in b2c12567de2d67c3577d007b679705da4706de43 .

**Review by stuartmorgan-g — APPROVED** [review]
LGTM with one comment about doc comments.

File: packages/flutter_tools/lib/src/flutter_plugins.dart
Comment (Gustl22) [reply packages/flutter_tools/lib/src/flutter_plugins.dart:233]: Nice catch, I mistakenly thought, that the check can be removed as every plugin should have a resolved implementation. But that's not checked for providing a default plugin. Should we warn the plugin author, if they reference a default plugin, which does not provide any implementations? I would not recommend to throw for compatibility reasons. I'll add some tests for that case.

Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/flutter_plugins.dart:233]: Logging a warning seems reasonable in that case.

**Comment by christopherfujino** [comment]
@andrewkolos can you give final approval on this one?

**Comment by Gustl22** [comment]
I suspect the CI failure comes from the red tree. Although it prints the warnings for the old gallery test correctly for having a missing plugin implementation for macos for the `connectivity` plugin.

File: packages/flutter_tools/lib/src/flutter_plugins.dart
Comment (Gustl22) [reply packages/flutter_tools/lib/src/flutter_plugins.dart]: I added some checks and tests in c076fc6008d9bd7efd2339c005594d5b08a3d1b6 and changed that comment in bc6cee3ac48d39d725f36efe4025f359b018e211.

Comment (Gustl22) [reply packages/flutter_tools/lib/src/flutter_plugins.dart]: Unfortunately I haven't found a clean solution to separate dart and native plugins for providing a warning, if no implementation is available for the default plugin. See dfd7991ce0dfacf1151aa1883c5f27e97c561da3 .
If that solution does not suit, I think I'll just remove it don't provide a warning in that case for now, as it is not directly related to this PR.

**Comment by Gustl22** [comment]
@stuartmorgan @andrewkolos can you afford some time for reviewing with the focus on how the [warning for missing implementation of default plugins](https://github.com/flutter/flutter/pull/137040#discussion_r1628238689) is provided. Thank you.

File: packages/flutter_tools/test/general.shard/dart_plugin_test.dart
Comment (stuartmorgan) [inline packages/flutter_tools/test/general.shard/dart_plugin_test.dart:821]: Is there a corresponding test that there's no warning when there's only a Dart implementation, and no native implementation? That's a valid case as well.

**Review by stuartmorgan-g — APPROVED** [review]
The changes still LGTM

**Comment by Gustl22** [comment]
I noticed by adding the latest test, that I completely ignored, that also Dart plugins are included in the `plugin.platforms` attribute. The implementation is not wrong, but the variable naming is misleading. So we only distinguish `Dart` and `DartOrNative`.

File: packages/flutter_tools/test/general.shard/dart_plugin_test.dart
Comment (Gustl22) [reply packages/flutter_tools/test/general.shard/dart_plugin_test.dart:821]: I added the test in 9c12ec84693fc56b93604b16275102cabcb32c76, but it fails, as I expect no native plugin is resolved, which is a false assumption. More [here](https://github.com/flutter/flutter/pull/137040#issuecomment-2178501864)

**Review by christopherfujino — APPROVED** [review]
LGTM (once tests pass, of course)

**Comment by Gustl22** [comment]
@stuartmorgan sorry for the delay and the continuous adaption. 
In the last three commits I refactored the code so it reflects the fact for distinguishing `Dart` and `nativeOrDart` plugin resolution (in contrast to having a `native` only resolution, which was stated wrong). Also adapted the test for this behavior, which originally was meant to test the default plugin warning only.

File: packages/flutter_tools/lib/src/flutter_plugins.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/flutter_plugins.dart:1332]: `dartpluginClass` wouldn't be correct, so the message here is misleading. It should say "pluginClass or dartPLuginClass".

File: packages/flutter_tools/lib/src/flutter_plugins.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/flutter_plugins.dart:110]: `nativeOrDart` doesn't seem like a plugin type, so this should probably be called `_PluginResolutionType`, and the comment adjusted accordingly.

**Review by stuartmorgan-g — APPROVED** [review]
Minor comments to the changes, but still LGTM with those adjustments.

File: packages/flutter_tools/lib/src/flutter_plugins.dart
Comment (Gustl22) [reply packages/flutter_tools/lib/src/flutter_plugins.dart]: Thx. Done in d621a904e4a2ae069a1d02926ce7142553ff6fe4

File: packages/flutter_tools/lib/src/flutter_plugins.dart
Comment (Gustl22) [reply packages/flutter_tools/lib/src/flutter_plugins.dart]: Done in d621a904e4a2ae069a1d02926ce7142553ff6fe4

**Comment by Gustl22** [comment]
@stuartmorgan Adapted your proposals. Thanks for the review.

**Comment by emakar** [comment]
Should I expect default implementation become absent once overridden?
I've added
```
  camera: 0.11.1
  camera_android: 0.10.10+1
```
but now pubspec.lock contains both `camera_android` and `camera_android_camerax`

**Comment by Gustl22** [comment]
As far as I remember, this is expected, as Flutter needs both packagages in order to check which one has implementations which can be overriden, when exclusively stating it as direct dependency.
Can you have a look if it' still written in `.flutter-plugins-dependencies` as dependency of `camera_android`?

**Comment by emakar** [comment]
I have both `camera_android` and `camera_android_camerax` among `dependencyGraph` in `.flutter-plugins-dependencies`

I actually don't understand why any changes were needed since we can just omit mention of "impl" plugin in `dependencies` of app facing plugin - then app can either choose impl 1, or impl 2, etc

I also don't understand why app facing plugin needs to define default plugin for endorsement whereas it could just depend on it effectively making it endorsed? (currently I see a common practice to both define a default impl and also depend on it)
Maybe I should read through documentation and flutter_plugins.dart once more :)

**Pull request merged**
